File: Vigenere/Vigenere.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

string Vigenere(string text, string key, bool encrypt);
int main()
{
    string text, key;
    bool encrypt;
    try
    {
        cout << "Enter text: ";
        getline(cin, text);
        cout << "Enter key: ";
        getline(cin, key);
        cout << "Encrypt or Decrypt? (1/0): ";
        cin >> encrypt;
        if (cin.fail())
        {
            throw invalid_argument("Invalid input for encryption choice.");
        }
        cout << "Result: " << Vigenere(text, key, encrypt) << endl;
    }
    catch (const exception &e)
    {
        cerr << "Error: " << e.what() << endl;
    }
    return 0;
}

vector<int> ConvertStringToInt(string str)
{
    vector<int> result;
    for (int i = 0; i < str.size(); i++)
    {
        if (isupper(str[i]))
        {
            result.push_back(str[i] - 'A');
        }
        else if (islower(str[i]))
        {
            result.push_back(str[i] - 'a');
        }
        else
            result.push_back(str[i]);
    }
    return result;
}

char ShiftChar(char c, int shift)
{
    if (shift < 0)
        shift += 26;
    if (isupper(c))
    {
        return (c - 'A' + shift) % 26 + 'A';
    }
    else if (islower(c))
    {
        return (c - 'a' + shift) % 26 + 'a';
    }
    return c;
}
string Vigenere(string text, string key, bool encrypt = true)
{
    vector<int> keyInt = ConvertStringToInt(key);
    string result = "";
    int keyIndex = 0;
    for (int i = 0; i < text.size(); i++)
    {
        if (keyIndex >= key.size())
            keyIndex = 0;
        result += ShiftChar(text[i], keyInt[keyIndex] * (encrypt ? 1 : -1));
        if (text[i] != ' ')
            keyIndex++;
    }
    return result;
}

File: Vigenere/Vigenere.py
def ConvertCharToInt(char):
    if char.islower():
        return ord(char) - ord('a')
    return ord(char) - ord('A')

def ChooseAlphabet(char, num):
    base = ord('a')
    if char.isupper():
        base = ord('A')
    return chr((ConvertCharToInt(char) + num)%26 + base)

def ShiftChar(char, shift, encrypt):
    if char.isalpha():
        shift = ConvertCharToInt(shift) * (1 if encrypt else -1)
        return ChooseAlphabet(char, shift)
    else:
        return char
    
def Vigenere(text, key, encrypt):
    result = ""
    keyIndex = 0
    for i in range(len(text)):
        if keyIndex >= len(key):
            keyIndex = 0
        result += ShiftChar(text[i], key[keyIndex], encrypt)
        if text[i] != ' ':
            keyIndex += 1
    return result

if __name__ == "__main__":
    try:
        text = input("Enter text: ")
        key = input("Enter key: ")
        encrypt = bool(int(input("Encrypt or Decrypt (1/0): ")))
    except ValueError:
        print("Invalid input. Please enter the correct values.")
        exit(1)
    print("Result: "+ Vigenere(text, key, encrypt))